Data/ByteString/Base64/Internal.hs

```diff
@@ -205,48 +206,104 @@ decodeWithTable padding decodeFP bs@(PS !fp !o !l) = unsafePerformIO $
-                  else if a .|. b .|. c .|. d == x
-                  then bail $ "invalid base64 encoding near offset " ++
-                       show (sp `minusPtr` sptr)
-                  else do
-                    poke8 dp $ fromIntegral (w `shiftR` 16)
-                    if c == done
-                      then finish $ n + 1
-                      else do
-                        poke8 (dp `plusPtr` 1) $ fromIntegral (w `shiftR` 8)
-                        if d == done
-                          then finish $! n + 2
-                          else do
-                            poke8 (dp `plusPtr` 2) $ fromIntegral w
-                            fill (dp `plusPtr` 3) (sp `plusPtr` 4) (n+3)
-          withForeignPtr dfp $ \dptr ->
-            fill dptr (sptr `plusPtr` soff) 0
+      withForeignPtr decodeFP $ \ !decptr ->
+        withForeignPtr sfp $ \sptr ->
+        withForeignPtr dfp $ \dptr -> do
+          let !end = sptr `plusPtr` (slen + soff)
+          decodeLoop
+            decptr
+            (plusPtr sptr soff)
+            dptr
+            end
+            dfp
+
+decodeLoop
+    :: Ptr Word8
+    -> Ptr Word8
+    -> Ptr Word8
+    -> Ptr Word8
+    -> ForeignPtr Word8
+    -> IO (Either String ByteString)
+decodeLoop !dtable !sptr !dptr !end !dfp = go dptr sptr 0
+  where
+    err p = return . Left
+      $ "invalid character at offset: "
+      ++ show (p `minusPtr` sptr)
+
+    padErr p =  return . Left
+      $ "invalid padding at offset: "
+      ++ show (p `minusPtr` sptr)
+
+    look :: Ptr Word8 -> IO Word32
+    look !p = do
+      !i <- peekByteOff p 0 :: IO Word8
+      !v <- peekByteOff dtable (fromIntegral i) :: IO Word8
+      return (fromIntegral v)
+
+    go !dst !src !n
+      | plusPtr src 4 >= end = finish dst src n
+      | otherwise = do
+        !a <- look src
+        !b <- look (src `plusPtr` 1)
+        !c <- look (src `plusPtr` 2)
+        !d <- look (src `plusPtr` 3)
+
+        if
+          | a == 0x63 -> padErr src
+          | b == 0x63 -> padErr (plusPtr src 1)
+          | c == 0x63 -> padErr (plusPtr src 2)
+          | d == 0x63 -> padErr (plusPtr src 3)
+          | a == 0xff -> err src
+          | b == 0xff -> err (plusPtr src 1)
+          | c == 0xff -> err (plusPtr src 2)
+          | d == 0xff -> err (plusPtr src 3)
+          | otherwise -> do
+
+            let !w = ((unsafeShiftL a 18)
+                  .|. (unsafeShiftL b 12)
+                  .|. (unsafeShiftL c 6)
+                  .|. d) :: Word32
+
+            poke8 dst (fromIntegral (unsafeShiftR w 16))
+            poke8 (plusPtr dst 1) (fromIntegral (unsafeShiftR w 8))
+            poke8 (plusPtr dst 2) (fromIntegral w)
+            go (plusPtr dst 3) (plusPtr src 4) (n + 3)
+
+    finish !dst !src !n
+      | src >= end = return (Right (PS dfp 0 n))
+      | otherwise = do
+        !a <- look src
+        !b <- look (src `plusPtr` 1)
+        !c <- look (src `plusPtr` 2)
+        !d <- look (src `plusPtr` 3)
```

Comment:
...doesn't this risk reading beyond the end of the `src` buffer? Most of the time this wouldn't matter, but in theory this might trigger segfaults in corner-cases....

...or maybe I'm just confused; so let me ask differently: which differences of `end - src` are possible on invocations of `finish`?

Reply:
No, because base64 encoded strings are gauranteed to have length `l = 0 mod 4`. We'll never have a situation where these will overflow. Inner loop iterates in chunks of 4 until we come to the last 4.